### PR TITLE
Initial Karma Support

### PR DESCRIPTION
**Pull request by ggetz**
Added Karma support as a first step for integrating tests as part of Travis-CI build process.

- Added Karma and necessary karma plugins to `package.json`, including launchers for:
 - Chrome
 - Firefox
 - IE
 - Safari
- Configured Karma to run jasmine tests
- Added `test` and `test-all` targets to `gulpfile.js`
 - `test` runs karma tests using browser specified in config file, currently Chrome
 - `test-all` runs karma tests using all browsers available

__TODO__
* [x] Add back in minified release and webGL validation support.
* [x] Update wiki for new test targets (categories and the two above)

__Post merge__
See #3571

### CODE REVIEW

**Comment by pjcozzi** [comment]
Exciting!  Can you please update any of the related Contributor's Guides in this branch?

**Comment by ggetz** [comment]
I added the new build targets to the Build Guide, and added a section on Karma in the Testing Guide.

**Comment by pjcozzi** [comment]
> I added the new build targets to the Build Guide, and added a section on Karma in the Testing Guide.

The updates look good, thanks!

File: gulpfile.js [hunk 1]
```diff
@@ -27,6 +28,26 @@ if (/\.0$/.test(version)) {
     version = version.substring(0, version.length - 2);
 }
 
+var karmaConfigFile = path.join(__dirname, 'Specs/karma.conf.js');
+
+//Manually patch list of karma binary files until bug https://github.com/karma-runner/karma/issues/1070 is addressed
```
Comment (mramato) [inline gulpfile.js]: I opened a pull request into karma to address the binary file issue: https://github.com/karma-runner/karma/pull/1816

Comment (ggetz) [inline gulpfile.js]: @mramato Awesome, thanks!

**Comment by mramato** [comment]
@ggetz, we should probably add back category support before this goes into master.  Can you branch off this and open a PR into this PR with the feature.  That should also enable us to run non-webgl tests on travis.

**Comment by ggetz** [comment]
> we should probably add back category support before this goes into master. Can you branch off this and open a PR into this PR with the feature. That should also enable us to run non-webgl tests on travis.

Yes

**Comment by ggetz** [comment]
I opened a PR to this branch at https://github.com/ggetz/cesium/pull/2

**Comment by mramato** [comment]
@ggetz can you merge in master when you get a chance?  Thanks.

**Comment by ggetz** [comment]
Merged.

**Comment by mramato** [comment]
I added some todos to the original PR comment.  Can other devs test this out and chime in with there thoughts.  I don't want to merge this and let it languish, I would like to move completely to karma ASAP, but it's going to take more than one PR.

One question up for debate is whether we care about the current HTML reporter.  I imagine lots of people like easily clicking on and running an individual test, so we might want to keep it.  However, karma also lets you run a single test by name or regex on the command line, so I could see that become the new way we handle those cases.

Please provide any feedback by __Monday, 2/8__.  @ggetz work on the main TODOs in this branch, but other stuff can be done in separate branches and we'll merge them afterwards.

**Comment by ggetz** [comment]
> @ggetz work on the main TODOs in this branch, but other stuff can be done in separate branches and we'll merge them afterwards.

OK

**Comment by pjcozzi** [comment]
Is this test supposed to fail?

image

**Comment by pjcozzi** [comment]
If we went without the HTML test runner, how do I use the debugger?  Do we need a debugger statement to initially launch the debugger?

**Comment by pjcozzi** [comment]
Would it be better for the debug config to not output every test to the console, and only output failures?

File: Documentation/Contributors/BuildGuide/README.md [hunk 2]
```diff
@@ -91,4 +91,6 @@ Here's the full set of scripts and what they do.
    * `clean` - Removes all generated build artifacts.
    * `cloc` - Runs [CLOC](https://github.com/AlDanial/cloc) to count the lines of code on the Source and Specs directories.  This requires [Perl](http://www.perl.org/) to execute.
    * `sortRequires` - Alphabetically sorts the list of required modules in every `js` file.  It also makes sure that the top of every source file uses the same formatting.
+   * `test` - Runs all tests with [Karma](http://karma-runner.github.io/0.13/index.html) using the default browser specified in the Karma config file.
+   * `test-all` - Runs all tests with [Karma](http://karma-runner.github.io/0.13/index.html) using all browsers installed on the current system.
```
Comment (pjcozzi) [inline Documentation/Contributors/BuildGuide/README.md]: The build targets are inconsistently named, e.g., `test-all` vs. `buildApps`.  Instead of `testAll`/`buildApps` or `test-all`/`build-apps`.

Could be a separate PR.

**Comment by pjcozzi** [comment]
Is it possible to launch the browser windows minimized?  Currently, this feels less frictionless than our current tests where I can just refresh and move to another window/tab.  Here there is an extra Window in the mix.

**Comment by pjcozzi** [comment]
As for the HTML reporter, I suspect we should keep it given how smooth the workflow is.  How hard is it to keep?

**Comment by pjcozzi** [comment]
Also we might want to consider ,merging this into a branch off of master so that we can try it _in practice_ on some of our own branches (and then just cherry-pick out our commits for PRs) for a few days.

**Comment by ggetz** [comment]
> If we went without the HTML test runner, how do I use the debugger? Do we need a debugger statement to initially launch the debugger?

> Would it be better for the debug config to not output every test to the console, and only output failures?

These are things that individual devs can tweak in their karma config file. You can configure `specReporter` options to change what is outputted (successes,fails,skips, and an error summary). For debugging you can disable the `singleRun` option and click debug mode in the browser.

**Comment by ggetz** [comment]
> Is it possible to launch the browser windows minimized? Currently, this feels less frictionless than our current tests where I can just refresh and move to another window/tab. Here there is an extra Window in the mix.

It looks like you can add flags to launch chrome minimized, but it is not available with the Firefox launcher.

**Comment by ggetz** [comment]
> As for the HTML reporter, I suspect we should keep it given how smooth the workflow is. How hard is it to keep?

We just would not remove it.

**Comment by pjcozzi** [comment]
>> Would it be better for the debug config to not output every test to the console, and only output failures?

> These are things that individual devs can tweak in their karma config file. You can configure specReporter options to change what is outputted (successes,fails,skips, and an error summary)

OK, but are we using the most sensible defaults for the core team?

**Comment by pjcozzi** [comment]
> It looks like you can add flags to launch chrome minimized, but it is not available with the Firefox launcher.

Given that we do most of our testing in Chrome, it sounds like a win.

**Comment by pjcozzi** [comment]
>> As for the HTML reporter, I suspect we should keep it given how smooth the workflow is. How hard is it to keep?

> We just would not remove it.

Let's keep it.  We can remove it later if we find ourselves not using it anymore.

File: Documentation/Contributors/BuildGuide/README.md (hunk 2 above)
Comment (mramato) [inline Documentation/Contributors/BuildGuide/README.md]: We should probably add a simple PR that does nothing by rename everything to be consistent, some of these have been a long time coming.  Then we can update this PR add new ones with the same style.  I'll open a PR with a proposal.

**Comment by mramato** [comment]
> As for the HTML reporter, I suspect we should keep it given how smooth the workflow is. How hard is it to keep?

I don't think it's hard to keep, but I think it will become less useful as people adopt to the new system.  My main complaint with the current reporter is that it loads slow, especially on mobile.

**Comment by mramato** [comment]
> Also we might want to consider ,merging this into a branch off of master so that we can try it in practice on some of our own branches (and then just cherry-pick out our commits for PRs) for a few days.

This is exactly what we are doing, except it can go into master and live in parallel to the old system.  I don't see a reason to make a separate branch since they can live side by side.  Do you disagree?

**Comment by pjcozzi** [comment]
> This is exactly what we are doing, except it can go into master and live in parallel to the old system. I don't see a reason to make a separate branch since they can live side by side. Do you disagree?

No, this is good.  I thought we had to remove the HTML Reporter.

**Comment by mramato** [comment]
> These are things that individual devs can tweak in their karma config file.

We should avoid having devs having to edit the karma config.  Everything can be done via the command line (or WebStorm).  I think our defaults are fine personally, I like seeing all of the tests run (and there's a summary of failures at the end).  However, when we get closer to merging this (and hopefully get some additional feedback from other devs) we can continue to tweak things.

**Comment by mramato** [comment]
> Is it possible to launch the browser windows minimized? Currently, this feels less frictionless than our current tests where I can just refresh and move to another window/tab. Here there is an extra Window in the mix.

I think this is an area we can improve things.  Right now we run the tests and then make the browser go away.  We can make the default keep the browser around and then you can just refresh the page (just like we do now).  In an ideal world there would be no browser at all and we would use phantomjs, but there are WebGL issues there.  As is always the case, Cesium pushes any tool to it's limit and there will be specific things to address that we haven't run into on other projects using karma.

**Comment by mramato** [comment]
> It looks like you can add flags to launch chrome minimized, but it is not available with the Firefox launcher.

This is probably fine, since Chrome is our default browser, but if we go with the "leave browser around for refresh", then it might not matter.

**Comment by mramato** [comment]
> If we went without the HTML test runner, how do I use the debugger? Do we need a debugger statement to initially launch the debugger?

I often go the debugger statement route even in our current set up.  But there are other options. If we don't close the browser across runs, you would just set your breakpoint and hit refresh.

On a related note, karma has the ability to run a specific test or set of tests via the command line, something like `karma run -- --grep=Cartesian3`, but I have to look into the exact details.  In theory if you are working on a single test, you could (via webstorm) have that one test set up to auto-run after every code change, decreasing the amount of switching between IDE and browser (since the result will get spit out in the IDE).

**Comment by mramato** [comment]
> No, this is good. I thought we had to remove the HTML Reporter.

Nope, we aren't removing anything until everyone is happy with the new system and any kinks are worked out.  There is also still the matter of coverage, which will be a huge improvement when we get it working.

**Comment by ggetz** [comment]
For all the different test supports, like categories, webgl validation, release - Should I be making individual gulp targets for each, or allowing for for taking this as arguments via the command line?

**Comment by mramato** [comment]
Good questions.  Given you may want to run them with different combinations, command line is probably the better way to go.  So it would be a single gulp task and then we can add script tags to package.json for the most common ones.

**Comment by ggetz** [comment]
> Good questions. Given you may want to run them with different combinations, command line is probably the better way to go. So it would be a single gulp task and then we can add script tags to package.json for the most common ones.

Sounds good, thanks

**Comment by ggetz** [comment]
I combined all the test tasks into one gulp task `test`. 

I used the NPM module [minimist](https://www.npmjs.com/package/minimist) for the command line args.

 I added the options `--all` to run all browsers available, `--webglValidation` to run with validation, and `--include [categoryName]` and `--exclude [categoryName]` for category support. I left the npm script names as is.

**Comment by mramato** [comment]
> I used the NPM module minimist for the command line args.

We already depend on `yargs` in Cesium.  They are similar, but I would use that instead. Sorry!

**Comment by ggetz** [comment]
> We already depend on yargs in Cesium. They are similar, but I would use that instead. Sorry!

OK, no problem

**Comment by ggetz** [comment]
I swapped out minimist for Yargs.

**Comment by ggetz** [comment]
I added the minified release support. I configured the require statements in `karma-main.js` similar to what was in `spec-main.js`. 

I think the release tests should be broken out into a separate task depending on `combine` and `minifyRelease`, rather than `build`, correct? We don't need to run `release` because we don't require the documentation which takes a long time to build.

**Comment by mramato** [comment]
I think I know what you're saying and I think I agree.  Basically turn your current test task into a function and have the tasks call it with the desired parameters.  Then have a different task for release tests that depends on `minifyRelease` instead of `build`.

**Comment by ggetz** [comment]
OK, thanks for understand what I was saying. I pulled out a `test` function and made a new task `test-release` that depends on `minifyRelease`.

**Comment by ggetz** [comment]
@mramato I've finished both the main todo's for this branch.

File: gulpfile.js [hunk 3]
```diff
@@ -274,6 +296,60 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
...
+    }
+
+    if (argv.webglValidation) {
+        webglValidation = true;
+    }
+
+    test(enableAllBrowsers, includeCategory, excludeCategory, webglValidation, false, done);
+});
+
+gulp.task('test-release', ['minifyRelease'], function(done){
```
Comment (mramato) [inline gulpfile.js]: Sorry for not realizing this until now, but it just occurred to me that we don't want to have and of the `test` tasks depend on anything.  The problem is that we often want to run the tests without building (for example, using the already produced release zip).  This means you could probably go back to a single test task that takes parameters, but that's up to you, whichever you think is cleaner.  Sorry again for the confusion.

File: package.json [hunk 4]
```diff
@@ -44,13 +44,23 @@
     "gulp-rename": "1.2.2",
     "gulp-replace": "0.5.4",
     "gulp-zip": "3.1.0",
+    "jasmine-core": "2.4.1",
     "jsdoc": "3.4.0",
     "jshint": "2.9.1",
     "jshint-stylish": "2.1.0",
+    "karma": "0.13.15",
```
Comment (mramato) [inline package.json]: A lot of these node modules are really out of date. To fix this, run `npm install -g npm-check-updates` on your machine and then run `ncu` inside of the building directory. It will inform you of all packages that need updating. Then you can run `ncu -u` to actually do the update. Then it's just a matter of npm install and testing that things still work.

Do not update `strip-comments` or `gulp-zip`, since both of those have outstanding issues with newer version.  You'll want to merge master into this branch first since there are already some packages that have been updated.

**Comment by ggetz** [comment]
I fixed the gulp task and updated the node modules. 

Everything seems to still be working, but the ` upsampled terrainData is copied to the tile once it is available` spec is still timing out. I've tried setting `jasmine.DEFAULT_TIMEOUT_INTERVAL` but that doesn't fix the issue.

**Comment by mramato** [comment]
Weird, I was not getting this error previously.  I'll take a look at it.

**Comment by pjcozzi** [comment]
What's left before we can merge this?  We can push as much as we want (except for branch deployment to S3 for testing) to a roadmap.  I just want to tie up loose ends so we can focus on 3D Tiles styling.

**Comment by mramato** [comment]
> except for branch deployment to S3 for testing

I don't think this has come up before?  What exactly is your goal here?

**Comment by mramato** [comment]
To answer your initial question, the bare minimum is done, They also just released a new Karma with my fix (thus removing the need for my hack), so I'm going to update to that.  Other than that, I just need to run through the release process and make sure we cover everything.

That being said, we should really get coverage switched over as well so that we can remove the old testing processes.  The longer we leave the migration half-done, the more painful it will be to maintain.

**Comment by pjcozzi** [comment]
S3 is for testing branches without having the pull them locally.  I'm fine with pushing coverage to the roadmap.

**Comment by mramato** [comment]
> S3 is for testing branches without having the pull them locally. I'm fine with pushing coverage to the roadmap.

That's not going to be possible with karma.  Plus that's strictly a CI thing.

**Comment by pjcozzi** [comment]
Why wouldn't we be able to, for example, run Sandcastle from a built Cesium branch hosted on S3?

**Comment by mramato** [comment]
Sorry, things like Sandcastle and Viewer will work, but actually running the tests won't.  But again, that has nothing to do with this PR, since these are karma/testing changes and have nothing to do with CI yet.

**Comment by mramato** [comment]
@ggetz I found an issue where categories do not work for tests at the `describe` level but only at the module level.  For example, `GlobeSurfaceTile` gets included in non-webgl tests when it should be excluded.  The current tests handle this properly so something is missing from the ported code.  Can you take a look?

**Comment by pjcozzi** [comment]
I guess I don't understand why we wouldn't be able to run the tests in a browser like we do today.

Regardless, let's put the CI items on a roadmap and evaluate what we can do right away and what can wait; I have to believe that this is really close to being CI ready.

**Comment by mramato** [comment]
> I have to believe that this is really close to being CI ready.

Absolutely, in fact we should be able to easily (in a new PR after this) have non-webgl tests run (both in AMD and "release" form) as part of the build.  We should also easily be able to upload the entire post-build source tree to S3 as a post-build step.

**Comment by ggetz** [comment]
> I found an issue where categories do not work for tests at the describe level but only at the module level. For example, GlobeSurfaceTile gets included in non-webgl tests when it should be excluded. The current tests handle this properly so something is missing from the ported code. Can you take a look?

OK, I'll take a look.

**Comment by pjcozzi** [comment]
:+1:

**Comment by mramato** [comment]
We can't run the tests from s3 by default because `karma` has a server component.  `karma` is what drives the tests compared to the static HTML file we have now.  We can probably get back to that point if we put in the time to do so, but we should integrate it into the new pipeline as opposed to maintaining the old and separate process.

**Comment by mramato** [comment]
@ggetz the category issue is the last thing needed before we can merge this.  Do you think you have time today to take a look?

**Comment by ggetz** [comment]
@mramato Ok, I added category support for the `describe` function. Does it also need to work for the `it` function?

**Comment by mramato** [comment]
> Does it also need to work for the it function?

It looks like the original code just uses defineSuite and describe, so I think we're good.

**Comment by ggetz** [comment]
@mramato Also I didn't mean to do that merge, can I undo it somehow?

**Comment by ggetz** [comment]
> It looks like the original code just uses defineSuite and describe, so I think we're good.

OK

**Comment by mramato** [comment]
I'll stop by.

**Comment by mramato** [comment]
This is almost ready, I'm probably going to make some final tweaks (hopefully tonight) to better fit normal cesium dev workflow and then I'll merge.  Thanks @ggetz.

**Comment by ggetz** [comment]
Awesome, thanks @mramato

**Comment by mramato** [comment]
This PR just got more awesome.  Building on @ggetz great work, I added Electron as a launcher option and set it to run headless.  I then make it the default when running via WebStorm's own Karma runner.  This means you can now run the tests from WebStorm with no additional windows or pop-ups and WebStorm presents the test in an easy to read tree with per-test output logs and clickable callstacks when things go wrong.

@pjcozzi and @ggetz would you might syncing this branch and giving it one last try.  After running `npm install` and opening WebStorm, you'll have a `Run tests` option in the upper right:

image

Give it a try and let me know what you think.

**Comment by pjcozzi** [comment]
Works great.  `fit` is also great.  Is there anything like that for an entire suite we can add for our `defineSuite` since we (at least, me) often run tests for a suite at a time?

**Comment by pjcozzi** [comment]
Also, add this to the test guide if you didn't already.

**Comment by mramato** [comment]
`defineSuite` does not exist in Jasmine, I'm not sure of the history as to why Cesium has it's own thing here instead of just using Jasmine `describe` directly (it's probably worth looking into). That being said, adding `fdefineSuite` should be easy.

**Comment by ggetz** [comment]
This works great! Very cool!

**Comment by pjcozzi** [comment]
I think the `defineSuite` history is AMD-related, but you would know better than me.

**Comment by mramato** [comment]
Okay, `fdefineSuite` is in and I updated the guides to include both new options as well as a quick Webstorm testing guide.

I also added a `--browsers` option for specifying browsers on the command line, i.e. `npm run test-webgl --browsers Firefox,Safari` will run web-gl only tests against Firefox and Safari.

**Comment by mramato** [comment]
While I'm sure there are plenty of kinks to work out, I think this is a huge step up for our dev workflow and is ready for master.  Since I did a bunch of work with @ggetz, I probably shouldn't be the one to hit merge; so @pjcozzi please do the honors when you are ready.

**Comment by pjcozzi** [comment]
:+1:

**Comment by mramato** [comment]
@kring you are no longer a liar.  You can thank @ggetz for it.

Thanks again @ggetz

**Comment by ggetz** [comment]
Thanks @mramato !

**Comment by kring** [comment]
Yay!  Thanks @ggetz! :)
